### PR TITLE
LVRUBY-112: Create model Project

### PR DESCRIPTION
**Pull request by kmandziy**
## JIRA ticket

_Please provide the URL for the JIRA ticket._

## Code reviewers

_Please list your code reviewers here._

## Summary of issue

CRUD for Projects
Ability to add Project Participant

## Summary of change

model Project
model ProjectParticipant
views for projects and project participants
Screenshot from 2019-05-22 17-48-28
Screenshot from 2019-05-22 18-07-28
Screenshot from 2019-05-24 02-47-46

## Review checklist

- [ ] Rubocop doesn't raise any warnings and offenses
- [ ] New features are covered with new specs
- [ ] All specs pass

### CODE REVIEW

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.48%** covered at **4.57** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.48%** covered at **4.61** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.79%** covered at **4.2** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.71%** covered at **4.07** hits/line

File: app/views/projects/_project_participants.html.haml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+- @project.participants.each do |participant|
+  %p
+    = participant.user.first_name
+    %b
+      = participant.role.name
+    = link_to 'Remove participant', project_participant_path(@project, participant),
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/views/projects/_project_participants.html.haml]: Outdated

File: app/assets/stylesheets/pages/projects.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+body .projects-page {
+  background-color: lightgrey;
```
[the remark this answers is not in the document]
Comment (kmandziy) [reply app/assets/stylesheets/pages/projects.scss]: Outdated

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.7%** covered at **4.08** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.7%** covered at **4.08** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.66%** covered at **4.11** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.66%** covered at **4.11** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.67%** covered at **4.09** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.69%** covered at **4.06** hits/line

File: app/controllers/projects_controller.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+class ProjectsController < ApplicationController
+  before_action :find_projects, only: %i[edit update destroy]
+
+  def index
+    @projects = Project.all
+  end
+
+  def show
+    @project = Project.find(params[:id])
+    @project_creator = User.find(@project.created_by)
```
Comment (okohuch) [inline app/controllers/projects_controller.rb]: You can create an association and get the author simply `@project.author` and use without a variable

File: app/models/project.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class Project < ApplicationRecord
+  mount_uploader :image, ImageUploader
+  validates :name, presence: true, length: { maximum: 100 }
+  validates :description, presence: true
+  validates :created_by, presence: true
```
Comment (okohuch) [inline app/models/project.rb]: You can create an association `belongs_to` which validates presence by default
and then you will get the author simply `@project.author` and use without a variable

`belongs_to :author,  class_name: 'User', foreign_key: 'created_by'`

File: app/views/projects/show.html.haml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+                = @project.name
+              %p.mb-2
+                = @project.description
+              .justify-content-center.mt-4
+                .user-short-info
+                  .thumbnail-social-icon.mr-2
+                    = image_tag @project_creator.avatar_url
+                  %p.ml-2.mb-1
+                    %b
+                      = @project_creator.first_name
```
Comment (okohuch) [inline app/views/projects/show.html.haml]: `@project.author.first_name`

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.06%** covered at **4.06** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.06%** covered at **4.06** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.24%** covered at **3.89** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.24%** covered at **3.89** hits/line

File: app/models/project.rb (hunk 4 above)
Comment (okohuch) [reply app/models/project.rb]: As I mentioned here, you can remove this validation since belongs_to validates presence of the foreign key by default

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.23%** covered at **3.9** hits/line

File: app/views/project_participants/_show.html.haml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+- if notice.present?
+  .notice.mb-3.text-success
+    = notice
+- if alert.present?
+  .notice.mb-3.text-danger
+    = alert
+- project.project_participants.each do |project_participant|
+  %p
+    - if project_participant.valid?
```
Comment (okohuch) [inline app/views/project_participants/_show.html.haml]: what is the reason to invoke the validations in the view?

File: app/controllers/projects_controller.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+  def edit; end
+
+  def create
+    @project = Project.new(project_params.merge(created_by: current_user.id))
+
+    if @project.save
+      redirect_to @project, notice: t('.notice')
+    else
+      render 'new', alert: t('.alert')
```
Comment (omaik) [inline app/controllers/projects_controller.rb]: symbol instead of string please, and below

File: app/models/project.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class Project < ApplicationRecord
+  mount_uploader :image, ImageUploader
+  validates :name, presence: true, length: { maximum: 100 }
+  validates :description, presence: true
+
+  belongs_to :author, class_name: 'User', foreign_key: 'created_by'
```
Comment (omaik) [inline app/models/project.rb]: is there a specific reason for doing so?

File: db/migrate/20190520070402_create_project_participants.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class CreateProjectParticipants < ActiveRecord::Migration[5.2]
+  def change
+    create_table :project_participants do |t|
+      t.references :user, index: true, foreign_key: true
+      t.references :role, index: true, foreign_key: true
```
Comment (omaik) [inline db/migrate/20190520070402_create_project_participants.rb]: why do we need a reference for the role?

File: app/controllers/projects_controller.rb [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    end
+  end
+
+  private
+
+  def project_params
+    params.require(:project).permit(:name, :description, :image)
+  end
+
+  def find_projects
```
Comment (omaik) [inline app/controllers/projects_controller.rb]: `find_project`

File: app/controllers/projects_controller.rb [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    @project = Project.find(params[:id])
```
Comment (omaik) [inline app/controllers/projects_controller.rb]: You do not limit the scope of search to the current_user  projects, so any user will be able to edit any project

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.19%** covered at **3.68** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.19%** covered at **3.68** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.19%** covered at **3.68** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.2%** covered at **3.67** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.2%** covered at **3.67** hits/line

File: app/views/project_participants/_show.html.haml [hunk 12]
```diff
@@ -0,0 +1,14 @@
+- if notice.present?
```
Comment (omaik) [inline app/views/project_participants/_show.html.haml]: is this something different from what we have in `application.html.haml`?

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.2%** covered at **3.67** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.2%** covered at **3.67** hits/line

**Pull request closed without merging**